subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/transaction/CompileTransaction.java

```diff
@@ -0,0 +1,372 @@
+ */
+
+package org.gradle.api.internal.tasks.compile.incremental.transaction;
+
+import org.gradle.api.UncheckedIOException;
+import org.gradle.api.internal.file.FileOperations;
+import org.gradle.api.tasks.WorkResult;
+import org.gradle.api.tasks.WorkResults;
+import org.gradle.api.tasks.util.PatternSet;
+import org.gradle.internal.file.Deleter;
+import org.gradle.language.base.internal.tasks.StaleOutputCleaner;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BooleanSupplier;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
+
+/**
+ * A helper class to handle incremental compilation after failure: it makes moving files around easier and reverting state easier.
+ */
+public class CompileTransaction {
+
+    private final Deleter deleter;
+    private final FileOperations fileOperations;
+    private final File tempDir;
+    private final List<TransactionalDirectory> transactionalDirectories = new ArrayList<>();
+    private final AtomicInteger uniqueDirIdGenerator = new AtomicInteger();
+    private Predicate<Throwable> stashThrowablePredicate = t -> true;
+
+    private CompileTransaction(File tempDir, FileOperations fileOperations, Deleter deleter) {
+        this.tempDir = tempDir;
+        this.fileOperations = fileOperations;
+        this.deleter = deleter;
+    }
+
+    /**
+     * Creates a new transactional directory in a temp directory that can be used to move files to.
+     * Actual directory is created only when execute() method is called.
+     */
+    public CompileTransaction newTransactionalDirectory(Consumer<TransactionalDirectory> directoryConsumer) {
+        File directory = new File(tempDir, "dir-uniqueId" + uniqueDirIdGenerator.getAndIncrement());
+        TransactionalDirectory transactionalDirectory = new TransactionalDirectory(directory, fileOperations, deleter);
+        directoryConsumer.accept(transactionalDirectory);
+        transactionalDirectories.add(transactionalDirectory);
+        return this;
+    }
+
+    /**
+     * Creates and registers transactional directory only if condition is met. If it is not met directory will not be register. Also, any action set on directory won't run.
+     */
+    public CompileTransaction newTransactionalDirectory(boolean registerCondition, Consumer<TransactionalDirectory> directoryConsumer) {
+        return registerCondition ? newTransactionalDirectory(directoryConsumer) : this;
+    }
```

Comment:
This conditional logic does not really fit into the builder pattern. I think we should ditch the builder altogether.